File: module2/pairSum.js
/*
pair sum

Write a function, pairSum, that takes in an array and a target sum as arguments. The function should return an array containing a pair of indices whose elements sum to the given target. The indices returned must be unique.

Be sure to return the indices, not the elements themselves.

There is guaranteed to be one such pair that sums to the target.
test_00:

pairSum([3, 2, 5, 4, 1], 8); // -> [0, 2]

test_01:

pairSum([4, 7, 9, 2, 5, 1], 5); // -> [0, 5]

test_02:

pairSum([4, 7, 9, 2, 5, 1], 3); // -> [3, 5]
*/

// brute force
const pairSum = (numbers, targetSum) => {
  let store = {};

  for (let i = 0; i < numbers.length; i++) {
    let currNumber = numbers[i];
    if (currNumber in store) return [store[currNumber], i];
    let difference = targetSum - currNumber;
    store[difference] = i;
  }
};

File: module2/pair_sum.py
"""
pair sum

Write a function, pair_sum, that takes in a list and a target sum as arguments. The function should return a tuple containing a pair of indices whose elements sum to the given target. The indices returned must be unique.

Be sure to return the indices, not the elements themselves.

There is guaranteed to be one such pair that sums to the target.
test_00:

pair_sum([3, 2, 5, 4, 1], 8) # -> (0, 2)

test_01:

pair_sum([4, 7, 9, 2, 5, 1], 5) # -> (0, 5)

test_02:

pair_sum([4, 7, 9, 2, 5, 1], 3) # -> (3, 5)

Approach:
- use a hash map so we can solve this in O(n) time and O(n) space
"""

# brute force
def pair_sum(numbers, target_sum):
  store = {}
  for i, num in enumerate(numbers):
    if num in store:
      return (store[num], i)
    difference = target_sum - num
    store[difference] = i